File: converted_code/backend/src/main/java/com/zeriq/ai/repository/UserRepository.java
package com.zeriq.ai.repository;

import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;
import com.zeriq.ai.model.User;
import java.util.Optional;

@Repository
public interface UserRepository extends JpaRepository<User, Long> {
    Optional<User> findByEmail(String email);
    Optional<User> findByUsername(String username);
}

File: src/backend/repositories/UserRepository.ts
import { dbStore, User } from "../database/dbStore";

export class UserRepository {
  public findById(id: number): User | undefined {
    return dbStore.data.users.find(u => u.id === id);
  }

  public findByEmail(email: string): User | undefined {
    return dbStore.data.users.find(u => u.email.toLowerCase() === email.toLowerCase());
  }

  public findByUsername(username: string): User | undefined {
    return dbStore.data.users.find(u => u.username.toLowerCase() === username.toLowerCase());
  }

  public save(user: User): User {
    const existingIndex = dbStore.data.users.findIndex(u => u.id === user.id);
    if (existingIndex >= 0) {
      dbStore.data.users[existingIndex] = user;
    } else {
      if (user.id === 0) {
        user.id = dbStore.data.users.length > 0 ? Math.max(...dbStore.data.users.map(u => u.id)) + 1 : 1;
      }
      dbStore.data.users.push(user);
    }
    dbStore.save();
    return user;
  }
}
